### ISSUE TITLE
Getter method in project

### ISSUE THREAD

**Issue by SweetPotato0213** [nus-cs2030/2021-s2#206]
# Summary
Are we allowed to use getter methods like getId() or getTime() in server and customer class?

**Comment by OuaisBien** [comment]
yes, thats 100% legal.

**Comment by gableejh** [comment]
Don't think there are any restrictions on that

**Comment by ARPspoofing** [comment]
Yeap for sure!

**Comment by ck-109** [comment]
This will violate the tell-don't-ask principle bah, in terms of design dk if prof will penalize. As much as possible, each class should be self contained.

**Comment by karintang** [comment]
yes, we are allowed to use getters

**Comment by SrinivasG99** [comment]
Yes. It was discouraged by prof in the due to the tell don't ask principal, but strictly speaking it is allowed.

**Comment by Jeyxiang** [comment]
Yes as long as codecrunch does not terminate your code!

**Comment by kaiyuuu** [comment]
Hello @SweetPotato0213, yes we are allowed to use getters for this project, but should avoid it as we should try keep to the "tell dont ask" principle :)

**Comment by SweetPotato0213** [comment]
> Hello @SweetPotato0213, yes we are allowed to use getters for this project, but should avoid it as we should try keep to the "tell dont ask" principle :)

Then without getters for this project, how can we access the ID and time of each customer in order to perform the next action of this customer?

**Comment by bandytan** [comment]
> > Hello @SweetPotato0213, yes we are allowed to use getters for this project, but should avoid it as we should try keep to the "tell dont ask" principle :)
> 
> Then without getters for this project, how can we access the ID and time of each customer in order to perform the next action of this customer?

Hello @SweetPotato0213, we are allowed to use getters for this project. I think what @kaiyuuu is trying to say is we should not abuse getters to access data in other classes. 

For example, in the Circle example in lecture, to find the distance between 2 points, we should not implement getX() and getY() to access the coordinates of the points and calculate it in Circle class. Instead, we should implement a distanceTo() method in Point class and call x.distanceTo(y) in Circle class. This is what "Tell-don't-Ask" principle is about.

**Comment by JoelToh404** [comment]
Yes! code crunch allows it

**Comment by chamath2030** [comment]
yes it's not a problem, most importantly is it works.

on a side note, shoutout to my GitHubbers Gwajoon, Wentinggy, Keithlimhs, naozitios, kaiyuuu, itsjiaming and elton. Continue living it big in the realm of SoC, hope to see you guys in other modules in the future.

**Comment by Ryanng-zr** [comment]
Hi! Yes we are allowed to use them. However on a side note, I would advise against using getter methods for lists as it would allow other classes to modify the list from outside the class itself. Otherwise, it should be fine!

**Comment by Jjjetplane** [comment]
Yes code crunch does not reject getters

**Comment by arnav0401** [comment]
I think it is a good practice to use getters in your code! The 2 primarily constraints that Code church will immediately look for is :
1. variables not being private final
2. presence of cyclical dependency

However, there is no restriction on using getters or in fact even setters

**Comment by ydymegan** [comment]
Yup it is okay to use getters but I think it is better to only have the getter method for things you need instead of setting up getters for all the attributes

**Comment by sebin0817** [comment]
hi yes to uphold immutability, you need and should use the getter methods as all your attributes must be private final! all the ebst
